src/ripple/app/tx/impl/DeleteAccount.cpp

```diff
@@ -0,0 +1,337 @@
+{
+    AccountID const account {ctx.tx[sfAccount]};
+    AccountID const dst {ctx.tx[sfDestination]};
+
+    auto sleDst = ctx.view.read(keylet::account(dst));
+
+    if (!sleDst)
+        return tecNO_DST;
+
+    if ((*sleDst)[sfFlags] & lsfRequireDestTag && !ctx.tx[~sfDestinationTag])
+        return tecDST_TAG_NEEDED;
+
+    // Check whether the destination account requires deposit authorization.
+    if (ctx.view.rules().enabled(featureDepositAuth) &&
+        (sleDst->getFlags() & lsfDepositAuth))
+    {
+        if (! ctx.view.exists (keylet::depositPreauth (dst, account)))
+            return tecNO_PERMISSION;
+    }
+
+    auto sleAccount = ctx.view.read (keylet::account(account));
+    assert(sleAccount);
+    if (! sleAccount)
+        return terNO_ACCOUNT;
+
+    // We don't allow an account to be deleted if its sequence number
+    // is within 256 of the current ledger.  This prevents replay of old
+    // transactions if this account is resurrected after it is deleted.
+    //
+    // We look at the account's Sequence rather than the transaction's 
+    // Sequence in preparation for Tickets.
+    constexpr std::uint32_t seqDelta {255};
+    if ((*sleAccount)[sfSequence] + seqDelta > ctx.view.seq())
+        return tecTOO_SOON;
+
+    // Verify that the account does not own any objects that would prevent
+    // the account from being deleted.
+    Keylet const ownerDirKeylet {keylet::ownerDir (account)};
+    if (dirIsEmpty(ctx.view, ownerDirKeylet))
+        return tesSUCCESS;
+
+    std::shared_ptr<SLE const> sleDirNode {};
+    unsigned int uDirEntry {0};
+    uint256 dirEntry {beast::zero};
+
+    if (! cdirFirst (
+        ctx.view, ownerDirKeylet.key, sleDirNode, uDirEntry, dirEntry, ctx.j))
+            // Account has no directory at all.  Looks good.
+            return tesSUCCESS;
+
+    std::int32_t deletableDirEntryCount {0};
+    do
+    {
+        // Make sure any directory node types that we find are the kind
+        // we can delete.
+        Keylet const itemKeylet {ltCHILD, dirEntry};
+        auto sleItem = ctx.view.read (itemKeylet);
+        if (! sleItem)
+        {
+            // Directory node has an invalid index.  Bail out.
+            JLOG (ctx.j.fatal()) << "DeleteAccount: directory node in ledger " 
+                << ctx.view.seq() << " has index to object that is missing: "
+                << to_string (dirEntry);
+            return tefBAD_LEDGER;
+        }
+        
+        LedgerEntryType const nodeType {
+            safe_cast<LedgerEntryType>((*sleItem)[sfLedgerEntryType])};
+
+        bool found {false};
+        for (CanDelete const& deletable : deletables)
+        {
+            if (deletable.type == nodeType)
+            {
+                found = true;
+                break;
+            }
+        }
+        if (! found)
+            // There's at least one directory entry that we can't delete.  Stop.
+            return tecHAS_OBLIGATIONS;
```

Comment:
I'd refactor this into a function that takes the ledger entry type and returns the pointer to the deleter function (if available). It can be used here and in the function below. I'll provide a couple patches you can cherry pick if you want to do this (patch will be in the "summary" comment so all the patches are together).

This also lets us remove the `CanDelete` struct and `deletables` array

Reply:
Cherry-picked 442f3cab6480dfafb291c4352ace3157671e75d3.